protons/driver.py

```diff
@@ -754,6 +755,110 @@ def _cache_force(self, titration_group_index, titration_state_index):
+        -----
+        The GHMC integrator is an simtk.openmm.CustomIntegrator object that calculates the protocol work internally.
+        Although it appears to only take one propagation step per perturbation step (i.e. ghmc.step(1)), multiple
+        propagation steps can be made depending on how the GHMC integrator was initialized.
+
+        To ensure the NCMC protocol is time symmetric, it has the form
+            propagation --> perturbation --> propagation
+
+        Parameters
+        ----------
+        context : simtk.openmm.Context
+            The context in which NCMC will be performed.
+
+        titration_group_indices :
+            The indices of the titratable groups that will be perturbed
+
+        initial_titration_states :
+            The initial protonation state of the titration groups
+
+        final_titration_states :
+            The final protonation state of the titration groups
+
+        Returns
+        -------
+        work : float
+          the protocol work of the NCMC procedure in multiples of kT.
+        """
+        # Turn the barostat and center of mass remover off, otherwise they contribute to the work
+        if self.barostat is not None:
+            self.barostat.setFrequency(0)
+        if self.cm_remover is not None:
+            self.cm_remover.setFrequency(0)
+
+        #TODO: remove explicit calculation of work with `getEnergy`
+        ghmc = self.compound_integrator.getIntegrator(1)
+        ghmc.setGlobalVariableByName("ntrials", 0)  # Reset the internally accumulated work
+        ghmc.setGlobalVariableByName("naccept", 0)  # Reset the GHMC acceptance rate counter
+
+        # The "work" in the acceptance test has a contribution from the titratable group weights.
+        g_initial = 0
+        for titration_group_index, (titration_group, titration_state_index) in enumerate(zip(self.titrationGroups, self.titrationStates)):
+            titration_state = titration_group['titration_states'][titration_state_index]
+            g_initial += titration_state['g_k']
+
+        # PROPAGATION
+        ghmc.step(1)
+
+        # The slow way to calculate the work
+        #work = 0.0
+
+        for step in range(self.nsteps_per_trial):
+
+            # Get the fractional stage of the the protocol
+            titration_lambda = float(step + 1) / float(self.nsteps_per_trial)
+
+            # The slow way to calculate the work
+            #nrg_initial = context.getState(getEnergy=True).getPotentialEnergy()
+
+            # PERTURBATION
+            for titration_group_index in titration_group_indices:
+                self._update_forces(titration_group_index, final_titration_states[titration_group_index],
+                                    initial_titration_state_index=initial_titration_states[titration_group_index],
+                                    fractional_titration_state=titration_lambda, context=context)
+            for force_index, force in enumerate(self.forces_to_update):
```

Comment:
As per the previous comment, `_update_forces()` no longer calls  `updateParametersInContext`.